### PR TITLE
Improve taxation specs

### PR DESCRIPTION
**Pull request by mamhoff**
This extracts the spec improvements (and only those) from https://github.com/solidusio/solidus/pull/536. 

What I do is I create a sort-of real-world spec set-up with understandable country and zone names and attach creative tax rates to them, then expect the order/line items to be correctly adjusted. 

A number of specs are set to pending because the current tax implementation can't do them. A number of other specs are xited because they produce random failures. 

There is one fix included for some failures that occur with the MOSS cases: If the zone your customer is ordering from is inside the default zone, but not the default zone, the default zone's taxes should still apply. 

The global diff is, unfortunately, horrific - I basically replace an entire section of specs. Git sees that a little differently. If you go through the commits one by one, you'll have a much easier time reviewing.

### CODE REVIEW

File: core/app/models/spree/tax_rate.rb [hunk 1]
```diff
@@ -142,8 +142,8 @@ def potentially_applicable?(order_tax_zone)
       self.zone == order_tax_zone ||
       # If the rate's zone *contains* the order's tax zone, then it's applicable.
       self.zone.contains?(order_tax_zone) ||
-      # 1) The rate's zone is the default zone, then it's always applicable.
-      (self.included_in_price? && self.zone.default_tax)
+      # 1) The rate's zone is in the default zone, then it's always applicable.
+      (self.included_in_price? && self.zone.contains?(Spree::Zone.default_tax))
```
Comment (cbrunsdon) [inline core/app/models/spree/tax_rate.rb]: was the logic here wrong before?

**Comment by cbrunsdon** [comment]
Just one quick question on one commit, but this looks awesome, thanks @mamhoff for pulling these apart (and I imagine pulling teeth while building them).

Will do a serious run through in the morning but first pass these looks great. Will reduce a ton of noise from #536 as well.

File: core/app/models/spree/tax_rate.rb (hunk 1 above)
Comment (mamhoff) [inline core/app/models/spree/tax_rate.rb]: Yes. If your default zone is Germany, and you have a tax rate defined for Europe, and then you try to ship outside Europe, you want to be aware that there's a European tax rate in your price. 

I deliberately chose a test setup with hairy cases like this, and reverting this to the original breaks seven specs: https://gist.github.com/mamhoff/512dae674e2437133eec

Comment (jhawthorn) [inline core/app/models/spree/tax_rate.rb]: The comment says:

> The rate's zone is in the default zone

But the code is

```
self.zone.contains?(Spree::Zone.default_tax))
```

Which reads to me as "the default zone is contained by the rate's zone"

Comment (mamhoff) [inline core/app/models/spree/tax_rate.rb]: Yes, correct, the comment is not correct in this case.

There is this idea of a smallest possible zone (`Spree::Zone.match(address)`) produces just that, the smallest fitting zone for an address. Apparently, a default zone should also be like that (contain no other zones). 

I'll fix the comment.

Comment (jhawthorn) [inline core/app/models/spree/tax_rate.rb]: Thanks Martin. I've cloned this down, and am trying to wrap my head around the tax system. I really appreciate format of the new specs here.

Comment (mamhoff) [inline core/app/models/spree/tax_rate.rb]: There's some fun bits in the production code. If you have questions, shoot them at me, I'm reachable on slack.

**Comment by jhawthorn** [comment]
I've given the specs a more thorough read and am very :+1:

**Comment by mamhoff** [comment]
Rebased on https://github.com/solidusio/solidus/pull/657 in order to separate concerns.

**Comment by mamhoff** [comment]
Fixups for fixed `pending` specs.

**Comment by cbrunsdon** [comment]
Yea, this is good by me too. Can starting to see this go in an actual reasonable direction now, thanks a lot @mamhoff.

:+1:

**Pull request merged**